### PR TITLE
Let exceptions fly while bootstrapping

### PR DESCRIPTION
**Pull request by clxmstaab**
with this change we get a proper exception stacktrace on the CLI.

without this change a Exception thrown while bootstrapping will just emit its message, e.g.

```sh
$ php ../phpstan-src/bin/phpstan analyse  -c app/partner/phpstan.neon.dist
Required package "complex/rocket" is not installed: cannot detect its version
```

after this changes we get a proper stacktrace, like one would expect who is used to symfony-console applications, which means depending on the use of `-v` parameter you get more details about the actual error:

```sh
$ php ../phpstan-src/bin/phpstan analyse  -c app/partner/phpstan.neon.dist
Required package "complex/rocket" is not installed: cannot detect its version

In CommandHelper.php line 306:

  [PHPStan\Command\InceptionNotSuccessfulException]


analyse [--paths-file PATHS-FILE] [-c|--configuration CONFIGURATION] [-l|--level LEVEL] [--no-progress] [--debug] [-a|--autoload-file AUTOLOAD-FILE] [--error-format ERROR-FORMAT] [--memory-limit MEMORY-LIMIT] [
--xdebug] [--] [<paths>...]
```

adding `-v`:
```sh
$ php ../phpstan-src/bin/phpstan analyse -v -c app/partner/phpstan.neon.dist
Required package "complex/rocket" is not installed: cannot detect its version

In CommandHelper.php line 306:

  [PHPStan\Command\InceptionNotSuccessfulException]


Exception trace:
  at C:\dvl\Zend\DefaultWorkspace13\phpstan-src\src\Command\CommandHelper.php:306
 PHPStan\Command\CommandHelper::begin() at C:\dvl\Zend\DefaultWorkspace13\phpstan-src\src\Command\AnalyseCommand.php:89
 PHPStan\Command\AnalyseCommand->execute() at C:\dvl\Zend\DefaultWorkspace13\phpstan-src\vendor\symfony\console\Command\Command.php:255
 Symfony\Component\Console\Command\Command->run() at C:\dvl\Zend\DefaultWorkspace13\phpstan-src\vendor\symfony\console\Application.php:934
 Symfony\Component\Console\Application->doRunCommand() at C:\dvl\Zend\DefaultWorkspace13\phpstan-src\vendor\symfony\console\Application.php:273
 Symfony\Component\Console\Application->doRun() at C:\dvl\Zend\DefaultWorkspace13\phpstan-src\vendor\symfony\console\Application.php:149
 Symfony\Component\Console\Application->run() at C:\dvl\Zend\DefaultWorkspace13\phpstan-src\bin\phpstan:60
 {closure}() at C:\dvl\Zend\DefaultWorkspace13\phpstan-src\bin\phpstan:61

analyse [--paths-file PATHS-FILE] [-c|--configuration CONFIGURATION] [-l|--level LEVEL] [--no-progress] [--debug] [-a|--autoload-file AUTOLOAD-FILE] [--error-format ERROR-FORMAT] [--memory-limit MEMORY-LIMIT] [
--xdebug] [--] [<paths>...]
```

DX wise this is a major improvement and saves a lot of time while debugging the cause of a error (without the need to install/activate xdebug)

### CODE REVIEW

**Comment by clxmstaab** [comment]
with the 2nd commit we actually give symfony the information to tell us about the cause of the exception:

```
$ php ../phpstan-src/bin/phpstan analyse  -c app/partner/phpstan.neon.dist
Required package "complex/rocket" is not installed: cannot detect its version

In CommandHelper.php line 306:

  Required package "complex/rocket" is not installed: cannot detect its version


In Versions.php line 103:

  Required package "complex/rocket" is not installed: cannot detect its version


analyse [--paths-file PATHS-FILE] [-c|--configuration CONFIGURATION] [-l|--level LEVEL] [--no-progress] [--debug] [-a|--autoload-file AUTOLOAD-FILE] [--error-format ERROR-FORMAT] [--memory-limit MEMORY-LIMIT] [
--xdebug] [--] [<paths>...]
```

File: src/Command/AnalyseCommand.php [hunk 1]
```diff
@@ -77,21 +77,17 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
-				$output,
-				$paths,
-				$pathsFile,
-				$memoryLimit,
-				$autoloadFile,
-				$configuration,
-				$level,
-				$allowXdebug
-			);
-		} catch (\PHPStan\Command\InceptionNotSuccessfulException $e) {
```
Comment (clxmstaab) [inline src/Command/AnalyseCommand.php]: a Exception thrown within a command will automatically lead to a exit-code of `1`

File: src/Command/CommandHelper.php [hunk 2]
```diff
@@ -303,7 +303,7 @@ public static function begin(
 				})($bootstrapFile);
 			} catch (\Throwable $e) {
 				$errorOutput->writeln($e->getMessage());
```
Comment (clxmstaab) [inline src/Command/CommandHelper.php]: IMO, in cases where we throw `InceptionNotSuccessfulException` with a root-cause exception provided, we should not print our the message manually but let symfony-console do its magic.

(this would apply to a few more places within this file)

File: src/Command/AnalyseCommand.php (hunk 1 above)
Comment (ondrejmirtes) [reply src/Command/AnalyseCommand.php]: Hi, I disagree with this change. CommandHelper outputs the error message, InceptionNotSuccessfulException signals the AnalyseCommand to return the exit code 1.

File: src/Command/CommandHelper.php (hunk 2 above)
Comment (ondrejmirtes) [reply src/Command/CommandHelper.php]: We could print out the stack trace in this place manually. The same way Symfony Console does. Can you investigate what function should we call to do that?

File: src/Command/AnalyseCommand.php (hunk 1 above)
Comment (ondrejmirtes) [reply src/Command/AnalyseCommand.php]: This change would lead to a really chaotic output in most cases which isn't needed.

File: src/Command/CommandHelper.php (hunk 2 above)
Comment (staabm) [reply src/Command/CommandHelper.php]: there is a `Symfony\Component\Console\Application::renderException()`.. I will have a look

File: src/Command/AnalyseCommand.php (hunk 1 above)
Comment (staabm) [reply src/Command/AnalyseCommand.php]: > InceptionNotSuccessfulException signals the AnalyseCommand to return the exit code 1.

symfony will automatically use a exit-code of 1 when a un-handled exception occurs.
no need to do this here and interrupt the exception handling

Comment (staabm) [reply src/Command/AnalyseCommand.php]: > This change would lead to a really chaotic output in most cases which isn't needed.

IMO the cause of chaotic output is that the output is scattered accross CommandHelper and the actual commands

Comment (ondrejmirtes) [reply src/Command/AnalyseCommand.php]: But the problem is that I don't always want to output the thrown exception with stack trace. I want to present something more friendly to the user.

File: src/Command/CommandHelper.php [hunk 3]
```diff
@@ -129,7 +131,7 @@ public static function begin(
 			try {
 				$projectConfig = $loader->load($projectConfigFile, null);
 			} catch (\Nette\InvalidStateException | \Nette\FileNotFoundException $e) {
-				$errorOutput->writeln($e->getMessage());
+				$application->renderException($e);
```
Comment (ondrejmirtes) [inline src/Command/CommandHelper.php]: Why this change? The message is usually about which file wasn't found which is sufficient.

File: src/Command/AnalyseCommand.php (hunk 1 above)
Comment (clxmstaab) [reply src/Command/AnalyseCommand.php]: when letting the exception fly through we would get a error like

image

no stacktrace or similar involved.

you get the stacktrace etc. when adding `-v` to the command

**Comment by ondrejmirtes** [comment]
There are still many problems with this. Decided not to go with it.
